### PR TITLE
Image pruner: Determine protocol just once

### LINKED ISSUE #15291: add back validation (maybe?) for `disable registry url validation in prune`
This was removed in `disable registry url validation in prune`

reenable 

```
if _, err := url.Parse(o.RegistryUrlOverride); 
   err != nil { in pkg/cmd/admin/prune/images.go
```

### PR DESCRIPTION
**Pull request by miminar**
Determine the registry protocol once. Do not change to other protocol during the run. This will produce nicer output without unrelated protocol fallback errors.

A follow-up for #14114 and #14405
Resolves #15291

### CODE REVIEW

**Comment by miminar** [comment]
[test][testextended][extended:core(registry|imageapi|ImagePrune|ImageQuota)]

File: pkg/image/prune/prune.go [hunk 1]
```diff
@@ -1190,3 +1160,62 @@ func getName(obj runtime.Object) string {
...
+	switch {
+	case len(url.Scheme) > 0 && url.Scheme != "unset":
+		protos = []string{url.Scheme}
+	case allowInsecure || netutils.IsPrivateAddress(registry):
+		protos = []string{"https", "http"}
+	default:
+		protos = []string{"https"}
+	}
+	registry = url.Host
+	glog.V(4).Infof("url: %#+v", url)
```
Comment (soltysh) [inline pkg/image/prune/prune.go]: This looks like a development debug, can you make it more friendly?

File: pkg/image/prune/prune.go [hunk 2]
```diff
@@ -1190,3 +1160,62 @@ func getName(obj runtime.Object) string {
...
+
+	for i, proto := range protos {
+		switch i {
+		case 0:
+			glog.V(4).Infof("Trying protocol %s for the registry url %s", proto, registry)
+		default:
+			glog.V(4).Infof("Falling back to protocol %s for the registry url %s", proto, registry)
+		}
```
Comment (soltysh) [inline pkg/image/prune/prune.go]: This is bad, just log the fact, you'll figure out when looking at logs, that this is a retry. There's no point in hardcoding this switch.

File: pkg/image/prune/prune.go [hunk 3]
```diff
@@ -1190,3 +1160,62 @@ func getName(obj runtime.Object) string {
...
+		url.Scheme = proto
+		err := callback(*url)
+		if err == nil {
+			return url, nil
+		}
+
+		if err != nil {
+			// TODO: remove the first
```
Comment (soltysh) [inline pkg/image/prune/prune.go]: Leftover?

File: pkg/image/prune/prune.go [hunk 4]
```diff
@@ -1190,3 +1160,62 @@ func getName(obj runtime.Object) string {
 	}
 	return fmt.Sprintf("%s/%s", accessor.GetNamespace(), accessor.GetName())
 }
+
+// tryProtocolsWithRegistryURL runs given callback with different protocols until no error is returned. The
+// https protocol is the first attempt. If it fails and allowInsecure is true, http will be the next. Obtained
+// errors will be concatenated and returned.
+func tryProtocolsWithRegistryURL(registry string, allowInsecure bool, callback func(registryURL url.URL) error) (*url.URL, error) {
```
Comment (soltysh) [inline pkg/image/prune/prune.go]: `callback` here suggests an actual callback method, but it's not. It's the heart of this method, an action performed by this function. I'd suggest renaming the variable.

Comment (soltysh) [inline pkg/image/prune/prune.go]: `action` maybe?

**Comment by soltysh** [comment]
Although it looks like your changes are failing tests, so you need to address that one.

File: pkg/image/prune/prune.go [hunk 5]
```diff
@@ -180,10 +181,11 @@ type defaultRegistryPinger struct {
 	insecure bool
 }
 
-func (drp *defaultRegistryPinger) ping(registry string) error {
-	healthCheck := func(proto, registry string) error {
+func (drp *defaultRegistryPinger) ping(registry string) (string, error) {
+	registryURL, err := tryProtocolsWithRegistryURL(registry, drp.insecure, func(u url.URL) error {
 		// TODO: `/healthz` route is deprecated by `/`; remove it in future versions
```
Comment (miminar) [inline pkg/image/prune/prune.go]: We can. Resolved.

File: pkg/image/prune/prune.go (hunk 4 above)
Comment (miminar) [inline pkg/image/prune/prune.go]: Good suggesion. Changed.

File: pkg/image/prune/prune.go (hunk 1 above)
Comment (miminar) [inline pkg/image/prune/prune.go]: You're right. Ditched.

File: pkg/image/prune/prune.go (hunk 2 above)
Comment (miminar) [inline pkg/image/prune/prune.go]: As you wish. I liked it better this way though.

File: pkg/image/prune/prune.go (hunk 3 above)
Comment (miminar) [inline pkg/image/prune/prune.go]: Yep. Removed.

**Comment by miminar** [comment]
Tests should be passing now.

**Comment by soltysh** [comment]
LGTM!

**Comment by miminar** [comment]
One last unit test fix.

**Comment by miminar** [comment]
Flake #14129 re-[test]

**Comment by miminar** [comment]
@mfojtik merge please.

**Comment by miminar** [comment]
I reformatted logging messages again and made a small change to defaulting to insecure connection.

Since now internal docker registry URL may be a DNS, let's not default to insecure connection if `--registry-url` override is empty. As long as a local IP is used as internal URL, insecure fall-back will be allowed.

**Comment by miminar** [comment]
Fixed broken formatting of help message.

**Comment by miminar** [comment]
Test flake #15450. re-[test]

**Comment by miminar** [comment]
Extended to address #15291.

File: pkg/cmd/admin/prune/images.go [hunk 6]
```diff
@@ -92,11 +107,10 @@ type PruneImagesOptions struct {
...
 	ForceInsecure       bool
 
-	OSClient       client.Interface
-	KClient        kclientset.Interface
-	RegistryClient *http.Client
-	Out            io.Writer
-	Insecure       bool
+	ClientConfig *restclient.Config
+	OSClient     client.Interface
+	KClient      kclientset.Interface
```
Comment (mfojtik) [inline pkg/cmd/admin/prune/images.go]: nit: KubeClient

File: pkg/cmd/admin/prune/images.go [hunk 7]
```diff
@@ -457,46 +504,48 @@ func (p *describingManifestDeleter) DeleteManifest(registryClient *http.Client,
...
+	}
+
+	osClient, kClient, err := f.Clients()
+	if err != nil {
+		return nil, nil, err
+	}
+	return osClient, kClient, err
+}
+
+// getRegistryClients returns a registry client. Note that registryCABundle and registryInsecure=true are
```
Comment (mfojtik) [inline pkg/cmd/admin/prune/images.go]: getRegistryClient

File: pkg/cmd/admin/prune/images.go [hunk 8]
```diff
@@ -550,12 +599,46 @@ func getClients(f *clientcmd.Factory, registryCABundle string, registryInsecure
...
 	}
 
-	registryClient = &http.Client{
+	return &http.Client{
 		Transport: wrappedTransport,
-	}
+	}, nil
+}
+
+func validateRegistryURL(registryURL string) error {
```
Comment (mfojtik) [inline pkg/cmd/admin/prune/images.go]: wish we move this into some helper package

File: pkg/image/prune/helper.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+	Ping(registry string) (*url.URL, error)
+}
+
+// DefaultRegistryPinger implements RegistryPinger.
+type DefaultRegistryPinger struct {
+	Client   *http.Client
+	Insecure bool
+}
+
+func (drp *DefaultRegistryPinger) Ping(registry string) (*url.URL, error) {
```
Comment (mfojtik) [inline pkg/image/prune/helper.go]: godoc

**Comment by miminar** [comment]
Comments addressed.

**Comment by miminar** [comment]
Flake #13575 and #15432 
/retest

**Comment by miminar** [comment]
Flakes #13575, #15405 /retest

**Comment by miminar** [comment]
@fabianofranz may I kindly request your approval?

File: pkg/image/prune/helper.go [hunk 10]
```diff
@@ -0,0 +1,205 @@
+package prune
```
Comment (miminar) [inline pkg/image/prune/helper.go]: It's in another file.

**Comment by miminar** [comment]
Yum flakes #8571 and #10162. Flake #15432

**Comment by miminar** [comment]
Jenkins job not found.
/retest

**Comment by miminar** [comment]
Flakes #15432 and #15763

**Comment by miminar** [comment]
Flake #15432 again.
/retest

**Comment by miminar** [comment]
Rebased. @soltysh may I have your lgtm one more time?

**Pull request merged**